source/include/azure_iot_hub_client.h

```diff
@@ -0,0 +1,428 @@
+
+    const uint8_t * pucVersion;               /**< The pointer to the twin document version. */
+    uint16_t usVersionLength;                 /**< The length of the twin document version. */
+} AzureIoTHubClientTwinResponse_t;
+
+/**
+ * @brief Cloud message callback to be invoked when a cloud message is received in the call to AzureIoTHubClient_ProcessLoop().
+ *
+ */
+typedef void ( * AzureIoTHubClientCloudToDeviceMessageCallback_t ) ( struct AzureIoTHubClientCloudToDeviceRequest * pxMessage,
+                                                             void * pvContext );
+
+/**
+ * @brief Method callback to be invoked when a method request is received in the call to AzureIoTHubClient_ProcessLoop().
+ *
+ */
+typedef void ( * AzureIoTHubClientMethodCallback_t ) ( struct AzureIoTHubClientMethodRequest * pxMessage,
+                                                       void * pvContext );
+
+/**
+ * @brief Twin callback to be invoked when a twin message is received in the call to AzureIoTHubClient_ProcessLoop().
+ *
+ */
+typedef void ( * AzureIoTHubClientTwinCallback_t ) ( struct AzureIoTHubClientTwinResponse * pxMessage,
+                                                     void * pvContext );
+
+/* Receive context to be used internally to the processing of messages */
+typedef struct AzureIoTHubClientReceiveContext
+{
+    struct
+    {
+        uint16_t usState;
+        uint16_t usMqttSubPacketID;
+        uint32_t ( * pxProcessFunction )( struct AzureIoTHubClientReceiveContext * pxContext,
+                                          AzureIoTHubClientHandle_t xAzureIoTHubClientHandle,
+                                          void * pvPublishInfo );
+
+        void * pvCallbackContext;
+        union
+        {
+            AzureIoTHubClientCloudToDeviceMessageCallback_t xCloudToDeviceMessageCallback;
+            AzureIoTHubClientMethodCallback_t xMethodCallback;
+            AzureIoTHubClientTwinCallback_t xTwinCallback;
+        } callbacks;
+    } _internal;
+} AzureIoTHubClientReceiveContext_t;
+
+typedef struct AzureIoTHubClientOptions
+{
+    const uint8_t * pucModuleID;  /**< The module ID to use for this device. */
+    uint32_t ulModuleIDLength;    /**< The length of the module ID. */
+
+    const uint8_t * pucModelID;   /**< The model ID used to identify the capabilities of a device based on the Digital Twin document. */
+    uint32_t ulModelIDLength;     /**< The length of the model ID. */
+
+    const uint8_t * pucUserAgent; /**< The user agent to use for this device. */
+    uint32_t ulUserAgentLength;   /**< The length of the user agent. */
+} AzureIoTHubClientOptions_t;
+
+typedef struct AzureIoTHubClient
+{
+    struct
+    {
+        AzureIoTMQTT_t xMQTTContext;
+
+        uint8_t * pucAzureIoTHubClientScratchBuffer;
+        uint32_t ulAzureIoTHubClientScratchBufferLength;
+        az_iot_hub_client xAzureIoTHubClientCore;
+
+        const uint8_t * pucHostname;
+        uint32_t ulHostnameLength;
```

Comment:
are these really uint32_t or uint16_t?

Reply:
hostnames I think are max 255 so we could make it `uint16_t`. 
https://stackoverflow.com/questions/8724954/what-is-the-maximum-number-of-characters-for-a-host-name-in-unix

Reply:
changed for device id as well:
https://github.com/MicrosoftDocs/azure-docs/blob/master/includes/iot-hub-limits.md